_posts/2020-09-01-processing-java-stream.md

```diff
@@ -0,0 +1,128 @@
+
+    }
+```
+
+Po dodaniu do kodu metody collect, dostajemy zamierzony efekt i mimo, że stworzona z przetworzonych danych lista nie jest przypisana do zmiennej, to strumień miał powód aby się wykonać.
+Listę metod pośrednich oraz końcowych znajdziemy [tutaj](https://stackoverflow.com/questions/47688418/what-is-the-difference-between-intermediate-and-terminal-operations).
+
+```java
+    private static final List<String> LIST = Arrays.asList("aab", "aab", 
+                                                           "aac", "aac", 
+                                                           "bbb", "ccc", "ddd", 
+                                                           "eee", "fff", "ggg");
+
+    public static void main(String[] args) {
+        LIST.stream()
+                .filter(e -> e.startsWith("a"))
+                .map(String::toUpperCase)
+                .peek(System.out::println)
+                .collect(Collectors.toList());
+    }
+```
+
+
+# Kolejność przetwarzania
+Cechą, która charakteryzuje strumienie jako „leniwe” jest kolejność wywołania operacji.
+Kod prezentuje przykładowe przetwarzanie danych oparte o metodę filter oraz map.
+
+```java
+    private static final List<String> LIST = Arrays.asList("aab", "aab", 
+                                                           "aac", "aac", 
+                                                           "bbb", "ccc", "ddd");
+    public static void main(String[] args) {
+        List<String> a = LIST.stream()
+                .filter(e -> {
+                    System.out.println("Stream - filter: " + e);
+                    return e.startsWith("a");
+                })
+                .map(e -> {
+                    System.out.println("Stream - map:" + e);
+                    return e.toUpperCase();
+                })
+                .collect(Collectors.toList());
+    }
+```
+Z logów wyświetlonych w konsoli widzimy, że operacje wykonywane są horyzontalnie, to znaczy dla każdego elementu od początku do końca. Dzięki temu, sekwencja skończy się na pierwszym niespełnionym warunku.
+```
+Stream - filter: aab
+Stream - map:aab
+Stream - filter: aab
+Stream - map:aab
+Stream - filter: aac
+Stream - map:aac
+Stream - filter: aac
+Stream - map:aac
+Stream - filter: bbb
+Stream - filter: ccc
+Stream - filter: ddd
+```
+
+# Przetwarzanie równoległe
+Wartościowym mechanizmem, który został dodany do Java Stream API jest możliwość równoległego przetwarzania strumieni, co daje możliwość wykorzystania większej liczby rdzeni procesora do wykonania zadania.
+Jak widać na załączonym kodzie wystarczy, że skorzystamy z metody **parallelStream** zamiast **stream** i nasze dane zostaną odpowiednio podzielone i przetworzone równolegle z użyciem wielu wątków. Warto zwrócić uwagę na to, że samo dzielenie danych wymaga dodatkowej pracy, dlatego przetwarzanie równoległe mniejszej ilości danych może trwać dłużej niż zrobienie tego samego z wykorzystaniem jednego wątku.
```

Comment:
Jakoś brakuje mi informacji, że trzeba uważać na parallel streamy ze względu na kolejność przetwarzania i fakt, że przy domyślnych parametrach to korzysta pod spodem ze wspólnej puli wątków